### PR TITLE
fix: inject dependencies into List[Message] input for Agent and Team

### PR DESCRIPTION
**Pull request by rodrigocoliveira**
When using `add_dependencies_to_context=True` on Agent or Team, dependencies were only injected into the user message when input was a string. This fix ensures dependencies are also injected when:

- Input is List[Message] (used by AGUI interface)
- Input is a single Message object

Changes:
- Agent._get_run_messages(): Inject dependencies into last user message for List[Message] input
- Agent._aget_run_messages(): Same fix for async version
- Team._get_user_message(): Inject dependencies for List[Message] and Message input
- Team._aget_user_message(): Same fix for async version

Fixes: add_dependencies_to_context ignored with List[Message] input

 Summary                                                                                               
                                                                                                        
  Fixes add_dependencies_to_context=True being ignored when input is List[Message] (used by AGUI interface).
                                                                                                        
  Problem                                                                                               
                                                                                                        
  When using add_dependencies_to_context=True on the Agent or Team constructor, dependencies were only injected into the user message when input was a string. This caused dependencies to never be added when using AGUI (or any interface that passes List[Message] to the agent).
                                                                                                        
  Root Cause                                                                                            
                                                                                                        
  In _get_run_messages(), only the string input path calls _get_user_message() which handles dependency injection. The List[Message] input path directly appended messages without any dependency processing.
                                                                                                        
  Changes                                                                                               
                                                                                                        
  - Agent._get_run_messages(): Inject dependencies into last user message for List[Message] input       
  - Agent._aget_run_messages(): Same fix for async version                                              
  - Team._get_user_message(): Inject dependencies for List[Message] and Message input                   
  - Team._aget_user_message(): Same fix for async version                                               
                                                                                                        
  Affected Input Types                                                                                  
                                                                                                        
  | Input Type    | Before     | After                |                                                 
  |---------------|------------|----------------------|                                                 
  | str           | ✅ Works   | ✅ Works             |                                                 
  | List[Message] | ❌ Ignored | ✅ Fixed             |                                                 
  | Message       | ❌ Ignored | ✅ Fixed (Team only) |                                                 
                                                                                                        
  Test plan                                                                                             
                                                                                                        
  - Verify dependencies appear in <additional context> when calling agent with string input             
  - Verify dependencies appear in <additional context> when calling agent with List[Message] input      
  - Verify other messages in the list are not modified                                                  
  - Verify no injection when add_dependencies_to_context=False                                          
  - Verify no injection when dependencies is None        

(If applicable, issue number: #\_\_\_\_)

## Type of change

- [ x ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Improvement
- [ ] Model update
- [ ] Other:

---

## Checklist

- [ x ] Code complies with style guidelines
- [ ] Ran format/validation scripts (`./scripts/format.sh` and `./scripts/validate.sh`)
- [ x ] Self-review completed
- [ ] Documentation updated (comments, docstrings)
- [ ] Examples and guides: Relevant cookbook examples have been included or updated (if applicable)
- [ ] Tested in clean environment
- [ ] Tests added/updated (if applicable)

---

## Additional Notes

### CODE REVIEW

**Comment by manuhortet** [comment]
Hey @rodrigocoliveira, this looks good! 

Can we add some tests before we merge?

**Comment by rodrigocoliveira** [comment]
Hello guys, I did not understand the process of merging here. Why is it blocked? Do i need to make something?

And this PR seems to conflict maybe with this... https://github.com/agno-agi/agno/pull/5832/commits/dc757b3c5b2c3784d5a8b6407e6134ee3615e4ee How to properly review and merge ?

**Comment by rodrigocoliveira** [comment]
@manuhortet why is this open yet? can we merge?

**Comment by rodrigocoliveira** [comment]
@manuhortet Can we deploy this?

**Comment by rodrigocoliveira** [comment]
@manuhortet Added some fix from another PR... Please I really need to get this released

**Comment by jpic** [comment]
They said they want some unit tests I think, great job btw

**Comment by rodrigocoliveira** [comment]
# ✅ Unit Tests Added — `test_dependencies_context.py`

As requested by @manuhortet and @jpic, I've added comprehensive unit tests covering all scenarios from the original test plan.

## Test Results: **14/14 PASS** ✅

---

### 🔧 `TestDependenciesWithListMessageInput` (5 tests)
*Core fix: `List[Message]` input — the primary bug this PR addresses*

| # | Test | Purpose | Result |
|---|------|---------|--------|
| 1 | `test_dependencies_injected_into_last_user_message` | The primary bug — dependencies were silently ignored for `List[Message]` input. Sends 3 messages, asserts last user message has `<additional context>` with both dependency values. | ✅ |
| 2 | `test_other_messages_not_modified` | Ensures injection is targeted — only the last user message is modified, not earlier ones. | ✅ |
| 3 | `test_dependencies_with_single_user_message_in_list` | Edge case — list with only one user message. The reverse-loop should still find and inject it. | ✅ |
| 4 | `test_dependencies_with_multimodal_content` | Covers OpenAI-format multimodal messages (`[{"type": "text", ...}, {"type": "image_url", ...}]`). | ✅ |
| 5 | `test_no_crash_when_no_user_messages_in_list` | Safety edge case — if the list has no user-role messages, must not crash. | ✅ |

---

### 🔧 `TestDependenciesWithListDictInput` (2 tests)
*AGUI interface path — the real-world trigger*

| # | Test | Purpose | Result |
|---|------|---------|--------|
| 6 | `test_dependencies_injected_with_dict_messages` | The actual AGUI path — AGUI sends `List[Dict]` with "role" keys. This is the real-world trigger for the bug. | ✅ |
| 7 | `test_dict_input_no_injection_when_disabled` | Negative control for the dict path — flag disabled should skip injection. | ✅ |

---

### 🔧 `TestNoDependenciesWhenDisabled` (2 tests)
*Flag gating*

| # | Test | Purpose | Result |
|---|------|---------|--------|
| 8 | `test_no_injection_when_disabled_list_message` | Verifies `False` flag prevents injection even when dependencies exist. | ✅ |
| 9 | `test_no_injection_when_default_none` | The parameter defaults to `None`. Since the condition is `if add_dependencies_to_context and ...`, `None` is falsy. | ✅ |

---

### 🔧 `TestNoDependenciesWhenNoneOrEmpty` (2 tests)
*Dependencies data gating*

| # | Test | Purpose | Result |
|---|------|---------|--------|
| 10 | `test_no_injection_when_dependencies_none` | No injection when `run_context.dependencies` is `None`, even with flag enabled. | ✅ |
| 11 | `test_no_injection_when_dependencies_empty_dict` | `{}` is falsy in Python. The condition correctly skips empty dicts. | ✅ |

---

### 🔧 `TestDependenciesWithStringInput` (2 tests)
*Regression check for existing behavior*

| # | Test | Purpose | Result |
|---|------|---------|--------|
| 12 | `test_dependencies_injected_with_string_input` | Regression guard — string input already worked before this PR. Ensures the refactor didn't break it. | ✅ |
| 13 | `test_no_injection_string_input_when_disabled` | Regression guard for the negative case on string input. | ✅ |

---

### 🔧 `TestDependenciesContentFormat` (1 test)
*Serialization correctness*

| # | Test | Purpose | Result |
|---|------|---------|--------|
| 14 | `test_dependencies_serialized_as_json` | Verifies injected content is valid JSON wrapped in `<additional context>` tags, matching expected structure. | ✅ |

---

## Test Plan Coverage ✅

All items from the original test plan are now covered:

- [x] Verify dependencies appear when calling agent with string input
- [x] Verify dependencies appear when calling agent with `List[Message]` input
- [x] Verify other messages in the list are not modified
- [x] Verify no injection when `add_dependencies_to_context=False`
- [x] Verify no injection when `dependencies` is `None`

Ready for merge! 🚀

**Comment by kausmeows** [comment]
@claude review pls

File: libs/agno/agno/agent/_messages.py [hunk 1]
```diff
@@ -1321,22 +1321,65 @@ def get_run_messages(
...
     ):
+        # Convert all messages first
+        converted_messages: List[Message] = []
         for _m in input:
             if isinstance(_m, Message):
-                run_messages.messages.append(_m)
-                if run_messages.extra_messages is None:
-                    run_messages.extra_messages = []
-                run_messages.extra_messages.append(_m)
+                converted_messages.append(_m)
```
Comment (kausmeows) [inline libs/agno/agno/agent/_messages.py:1328]: thinking about this, i think we should pass a deepcopy of that message object to avoid mutating the state.

File: libs/agno/agno/agent/_messages.py [hunk 2]
```diff
@@ -1321,22 +1321,65 @@ def get_run_messages(
...
                     msg = Message.model_validate(_m)
-                    run_messages.messages.append(msg)
-                    if run_messages.extra_messages is None:
-                        run_messages.extra_messages = []
-                    run_messages.extra_messages.append(msg)
+                    converted_messages.append(msg)
                 except Exception as e:
                     log_warning(f"Failed to validate message: {e}")
 
+        # Inject dependencies into the last user message if enabled
```
Comment (kausmeows) [inline libs/agno/agno/agent/_messages.py:1336]: this whole new logic can perhaps be in a helper function instead of being used twice in sync/async versions

File: libs/agno/agno/agent/_messages.py [hunk 3]
```diff
@@ -1523,22 +1566,65 @@ async def aget_run_messages(
...
+        if add_dependencies_to_context and dependencies is not None:
+            dependencies_block = (
+                "\n\n<additional context>\n"
+                + convert_dependencies_to_string(agent, dependencies)
+                + "\n</additional context>"
+            )
+            for i in range(len(converted_messages) - 1, -1, -1):
+                msg = converted_messages[i]
+                if msg.role == "user":
+                    if isinstance(msg.content, str):
```
Comment (kausmeows) [inline libs/agno/agno/agent/_messages.py:1592]: what if content is None?
